teuthology/provision/openstack.py

```diff
@@ -57,52 +57,100 @@ def init_user_data(self, os_type, os_version):
+        json_result = misc.sh("openstack %s -f json" % subcommand)
+        r = json.loads(json_result)
+        if get:
+            return self.get_value(r, get)
+        return r
+
+    def _create_volume(self, volume_name, size):
+        """
+        Create a volume and return valume id
+        """
+        volume_id = None
+        try:
+            volume_id = self._openstack("volume show %s" % volume_name, 'id')
+        except subprocess.CalledProcessError as e:
+            if 'No volume with a name or ID' not in e.output:
+                raise e
+        if volume_id:
+            log.warn("Volume {} already exists with ID {}; using it"
+                     .format(volume_name, volume_id))
+        volume_id = self._openstack(
+            "volume create %s" % config['openstack'].get('volume-create','')
+            + " --property ownedby=%s" % config['openstack']['ip']
+            + " --size %s" % str(size) + ' ' + volume_name, 'id')
+        if volume_id:
+            log.info("Volume {} created with ID {}"
+                     .format(volume_name, volume_id))
+            return volume_id
+        else:
+            raise Exception("Failed to create volume %s" % volume_name)
+
+    def _await_volume_status(self, volume_id, status='available'):
+        """
+        Wait for volume to have status, like 'available' or 'in-use'
+        """
+        with safe_while(sleep=4, tries=50,
+                        action="volume " + volume_id) as proceed:
+            while proceed():
+                try:
+                    volume_status = \
+                        self._openstack("volume show %s" % volume_id, 'status')
+                    if volume_status == status:
+                        break
+                    else:
+                        log.debug("volume %s not in '%s' status yet"
+                                  % (volume_id, status))
+                except subprocess.CalledProcessError:
+                        log.warn("volume " + volume_id +
+                                 " not information available yet")
+
+    def _attach_volume(self, volume_id, name):
+        """
+        Attach volume to OpenStack instance.
+
+        Try and attach volume to server, wait until volume gets in-use state.
+        """
+        with safe_while(sleep=20, increment=20, tries=3,
+                        action="add volume " + volume_id) as proceed:
+            while proceed():
+                try:
+                    misc.sh("openstack server add volume " + name + " " + volume_id)
+                    break
+                except subprocess.CalledProcessError:
+                    log.warning("openstack add volume failed unexpectedly; retrying")
+        self._await_volume_status(volume_id, 'in-use')
+
+    def _remove_volume(self, volume_id):
+        OpenStack().run("volume set --name REMOVE-ME %s || true" % volume_id)
+        OpenStack().run("volume delete %s || true" % volume_id)
```

Comment:
nit: this is duplicating code from https://github.com/SUSE/teuthology/blob/master/teuthology/openstack/__init__.py#L193-L194 ?

I suggest to move `_remove_volume()` to a shared module and call it from both places.